### PR TITLE
first pass w/ expression engine

### PR DESCRIPTION
**Pull request by maxres-fr**
DRAFT

### CODE REVIEW

File: pkg/health/ctrl.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    - resource: secretagentconfigurations
+      name: forgerock-sac
+      apiversion: v1alpha1
+      group: secret-agent.secrets.forgerock.io
+      checks:
+        - expression: status.totalManagedObjects == 12
+          timeout: 0s
+        - expression: status.state == "Completed"
+          timeout: 0s
+`)
```
Comment (jrcast) [inline pkg/health/ctrl.go:28]: 👍

File: pkg/health/health.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		if err != nil {
+			return false, err
+		}
+		return output.(bool), nil
+	}
+}
+
+// Check run wait on resource until expression passes
+// note at the moment it doesn't track any scycess/fails
+func (h *Resource) Check(clientFactory factory.Factory, namespace string) (bool, error) {
```
Comment (jrcast) [inline pkg/health/health.go:54]:
```suggestion
func (r *Resource) Check(clientFactory factory.Factory, namespace string) (bool, error) {
```
r for resource?

File: pkg/health/health.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	clientMgr := k8s.NewK8sClientMgr(clientFactory)
+	namespace, err := clientMgr.Namespace()
+	if err != nil {
+		return false, err
+	}
```
Comment (jrcast) [inline pkg/health/health.go:59]: namespace is already passed in to the function. No need to get it again.
```suggestion
```

File: pkg/health/health.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+			err = errors.Wrapf(err, "%s checks failed", r.Name)
+
+		}
+		if !healthy {
+			h.unhealthy = append(h.unhealthy, r.Name)
+			continue
+		}
+		if healthy {
+			h.healthy = append(h.healthy, r.Name)
+		}
```
Comment (jrcast) [inline pkg/health/health.go:114]: don't need second `if` after `continue`
```suggestion
		if !healthy {
			h.unhealthy = append(h.unhealthy, r.Name)
			continue
		}
		h.healthy = append(h.healthy, r.Name)
		
```

**Review by jrcast — APPROVED** [review]

File: docs/forgeops_apply.md [hunk 5]
```diff
@@ -29,7 +29,7 @@ Deploy common platform components
 ```
       --as string                      Username to impersonate for the operation
       --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
-      --cache-dir string               Default cache directory (default "/home/jcastillo/.kube/cache")
+      --cache-dir string               Default cache directory (default "/home/max/.kube/cache")
```
Comment (jrcast) [inline docs/forgeops_apply.md:32]: we have something going on here. 🤷

Comment (maxres-fr) [reply docs/forgeops_apply.md:32]: yea this is annoying

File: pkg/health/health.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		if err != nil {
+			return false, err
+		}
+		return output.(bool), nil
+	}
+}
+
+// Check run wait on resource until expression passes
+// note at the moment it doesn't track any success/fails
+func (h *Resource) Check(clientMgr k8s.ClientMgr, namespace string) (bool, error) {
```
Comment (jrcast) [inline pkg/health/health.go:53]:
```suggestion
func (r *Resource) Check(clientMgr k8s.ClientMgr, namespace string) (bool, error) {
```

File: cmd/doctor.go [hunk 7]
```diff
@@ -34,15 +31,14 @@ var ds = &cobra.Command{
 		printer.Warnln("Not implemented")
 		return nil
 	},
-	SilenceUsage:      true,
```
[the remark this answers is not in the document]
Comment (jrcast) [reply cmd/doctor.go:37]: @maxres-fr 👀

File: cmd/doctor.go [hunk 8]
```diff
@@ -68,12 +59,12 @@ var ignoreOperators []string
...
 	Aliases: []string{"op"},
-	Short:   "Verify that operators are installed and ready",
+	Short:   "Check Operators Installed and Running",
 	Long: `
-    Verify that operators are installed and running:
-	* Checks to ensure that required operators are installed and ready.
-	* Searches all namespaces for the default deployment of secret agent, nginx-ingress, cert-manager.
-	* Checks for a minimum ready count of one.`,
+	Checks to ensure that required operators are installed and ready.
+	Searches all namespaces for the default deployment of secret agent, nginx-ingress, cert-manager
```
Comment (dgoldssfo) [inline cmd/doctor.go:65]: s/cert-manager/and cert-manager/

File: docs/forgeops_doctor.md [hunk 9]
```diff
@@ -34,6 +35,6 @@ Diagnose common cluster and platform deployments
 ### SEE ALSO
 
 * [forgeops](forgeops.md)	 - forgeops is a tool for managing ForgeRock platform deployments
-* [forgeops doctor operator](forgeops_doctor_operator.md)	 - Verify that operators are installed and ready
+* [forgeops doctor operator](forgeops_doctor_operator.md)	 - Check Operators Installed and Running
```
Comment (dgoldssfo) [inline docs/forgeops_doctor.md:38]: If this ends up in help text, revert it to what it was (but change "ready" to "running")

File: docs/forgeops_doctor_operator.md [hunk 10]
```diff
@@ -1,14 +1,14 @@
 ## forgeops doctor operator
 
-Verify that operators are installed and ready
+Check Operators Installed and Running
```
Comment (dgoldssfo) [inline docs/forgeops_doctor_operator.md:3]: Same here

File: pkg/health/ctrl.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+		return err
+	}
+	allHealthy, err := hlth.CheckResources(clientMgr)
+	if err != nil {
+		return err
+	}
+	if !allHealthy {
+		numHealthy := len(hlth.healthy)
+		totalNum := len(hlth.Spec.Resources)
+		printer.Warnf("health check %s has %d / %d healthy resources", hlth.Name, numHealthy, totalNum)
```
Comment (dgoldssfo) [inline pkg/health/ctrl.go:44]: Please start this (and other) printed messages with a capital letter. i.e.,

printer.Warnf("Health check %s has %d / %d healthy resources", hlth.Name, numHealthy, totalNum)

File: pkg/health/ctrl.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+		for _, resourceName := range hlth.healthy {
+			printer.Noticef("resource %s healthy", resourceName)
```
Comment (dgoldssfo) [inline pkg/health/ctrl.go:46]: printer.Noticef("Resource %s is healthy", resourceName)

File: pkg/health/ctrl.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+		}
+		for _, resourceName := range hlth.unhealthy {
+			printer.Warnf("resource %s not healthy", resourceName)
```
Comment (dgoldssfo) [inline pkg/health/ctrl.go:50]: printer.Warnf("Resource %s is not healthy", resourceName)

File: pkg/health/ctrl.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+		}
+		return nil
+	}
+	printer.Noticef("health check %s has passed", hlth.Metadata.Name)
```
Comment (dgoldssfo) [inline pkg/health/ctrl.go:54]: printer.Noticef("Health check %s has passed", hlth.Metadata.Name)

**Review by dgoldssfo — COMMENTED** [review]
Looks good, just some changes to text strings

**Comment by maxres-fr** [comment]
Missed some comments here, so I added them in PR#41

**Pull request merged**